src/derivative_utils.jl

```diff
@@ -392,6 +392,57 @@ function calc_W!(integrator, cache::OrdinaryDiffEqMutableCache, dtgamma, repeat_
   return nothing
 end
 
+function calc_W!(integrator, cache::OrdinaryDiffEqMutableCache, dtgamma, repeat_step, W_index::Int, W_transform=false)
+  @unpack t,dt,uprev,u,f,p = integrator
+  @unpack J,W = cache
```

Comment:
this is just silly. Why not just pass in `W` into these routines? It would fix these problems.

Reply:
@kanav99 @huanglangwen this is something we should follow up on. It's because our original `calc_W!` was never intended for thread-safety, and so to make it thread-safe there is this code duplication. We should assume this case can happen in the next refactor.

Reply:
Does it mean `calc_W!` should have an additional parameter of *readonly* `W` ?

Reply:
it should have `J and `W` I think, where it writes into `W` for a given `J`, instead of pulling them from the cache (since there may be more than 1)

Reply:
But it won't work on OOP.

Reply:
For OOP it will need to return W.

Reply:
I think we should have a pseudo-nlsolver kind of thing for algorithms which don't have an actual nlsolver like Rosenbrocks, the struct should have aliases to the `J` and `W` and similar stuff, should be `<: NLSolver` so that we have a common implementation of the functions. This way it won't make us change the original derivative utilities everytime we make a new algorithm just for an overload.

Reply:
Note that algorithms with nlsolver struct are already well handled for parallel applications.